include/swift/AST/Import.h

```diff
@@ -375,18 +387,53 @@ class ImportPath : public detail::ImportPathBase<ImportPath> {
   /// including submodules, assuming the \c ImportDecl has the indicated
   /// \c importKind.
   Module getModulePath(ImportKind importKind) const {
-    return getModulePath(importKind != ImportKind::Module);
+    return getModulePath(isScopedImportKind(importKind));
   }
 
   /// Extracts the portion of the \c ImportPath which represents a scope for the
   /// import, assuming the \c ImportDecl has the indicated \c importKind.
   Access getAccessPath(ImportKind importKind) const {
-    return getAccessPath(importKind != ImportKind::Module);
+    return getAccessPath(isScopedImportKind(importKind));
   }
 };
 
 // MARK: - Abstractions of imports
 
+/// Convenience struct to keep track of an import path and whether or not it
+/// is scoped.
+class UnloadedImportedModule {
```

Comment:
Since you pinged me about naming :P, `UnloadedImportedModule` to me conveys a sense of "the module was imported and loaded then unloaded", i.e. unloading as something that undoes loading. However, IIUC that's not how this is being used. Rather the semantics are more like `NotYetImportedModule` (strawman alternative). Is that right?

Reply:
I suspect part of the issue (for my lack of understanding) is that there are several moving parts here -- there's `UnboundImport` as well as `UnloadedImportedModule` (which kinda' sound similar-ish) -- and I only have passing familiarity with this code, so I don't fully grasp the distinction between loading modules and binding imports.

Reply:
Think of it this way. We have these three public types which are used throughout the compiler:

* `ImportedModule` is basically just a `ModuleDecl *` and an access path. It's the minimum information you need to, for instance, figure out if a particular declaration is visible through that import.

* `UnloadedImportedModule` is basically the *name* of a module and an access path. You use it in place of `ImportedModule` when you haven't actually loaded a `ModuleDecl *` yet.

* `AttributedImport<ImportedModule or UnloadedImportedModule>` adds to that type abstract information about the attributes of an import (like whether it's exported, implementation-only, which SPIs it imports, etc.).

And we have one other type that's private to ImportResolution.cpp:

* `UnboundImport` contains state and logic used to process a single import. (Could be an `ImportDecl`, a cross-import overlay we need to import, or now an implicit import.) In terms of state, it includes an `AttributedImport<UnloadedImportedModule>` plus a couple extra things that are used when emitting diagnostics. But its primary purpose is to perform certain steps of import resolution on a particular import, not to store state. (Arguably it should actually be `UnresolvedImport`; when I wrote this class, "import resolution" was called "name binding".)

Reply:
I don't mind the name `UnloadedImportedModule`, though perhaps another suggestion could be `ImportedModuleToLoad`?

Reply:
Does `ImportedModule` always represent the state where a `ModuleDecl *` has been loaded? Or does it sometimes have a `ModuleDecl *` that hasn't been loaded?

Reply:
ModuleDecls are, by definition, already loaded.